vmm/src/lib.rs

```diff
@@ -638,6 +640,54 @@ impl VmmVersionInfo {
     }
 }
 
+struct VmGuard {
+    vm: Option<Arc<Mutex<Vm>>>,
+    sender: std::sync::mpsc::Sender<Option<Vm>>,
+    is_local: bool,
+    check_migration_evt: Option<EventFd>,
+}
+
+impl Drop for VmGuard {
+    fn drop(&mut self) {
+        // Return the VM only if the migration fails
+        if let Some(vm) = self.vm.take() {
+            if let Ok(mutex) = Arc::try_unwrap(vm) {
+                if let Ok(mut inner_vm) = mutex.into_inner() {
```

Comment:
nit: if someone restarts the work (which I might do soon): here we have great potential for let chains - less nesting!